[JIT] Support scripting torch.is_autocast_enabled()

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81305

This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

Differential Revision: [D37901585](https://our.internmc.facebook.com/intern/diff/D37901585)

davidberard98 committed:
[JIT] Support scripting torch.is_autocast_enabled()

This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37901585).

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

Differential Revision: [D37901585](https://our.internmc.facebook.com/intern/diff/D37901585)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

Differential Revision: [D37901585](https://our.internmc.facebook.com/intern/diff/D37901585)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37901585).

eellison reviewed (commented):
I think this won't compose with the amp pass right now, can you add some tests where within script we turn on amp, and then call these functions?

davidberard98 commented:
hmm good catch @eellison, you're right that doesn't work.

I guess it's a bit tricky because in the following example:
```python
import torch

def fn(x, y):
    with torch.cpu.amp.autocast(enabled=False):
        return torch.mm(x, y)

x = y = torch.rand(2, 2)

fn_s = torch.jit.script(fn)

with torch.cpu.amp.autocast(enabled=True):
    print(fn_s(x, y).dtype)
```

we return `torch.bfloat16`... because JIT-autocast doesn't do the casting, but eager-autocast does turn on and therefore autocasting does occur

I guess maybe the right way to handle this is to return true if either of JIT autocast or eager autocast is turned on.

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

Differential Revision: [D37901585](https://our.internmc.facebook.com/intern/diff/D37901585)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

Differential Revision: [D37901585](https://our.internmc.facebook.com/intern/diff/D37901585)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37901585).

davidberard98 commented on torch/csrc/jit/passes/autocast.cpp:
```diff
@@ -216,6 +216,39 @@ void castInputsToWidestType(Node* node, const AutocastContext& context) {
+// invoked, then autocasting will occur regardless of what the JIT-autocasting
+// settings are.
+void updateAutocastEnabledCheck(Node* node, bool is_jit_enabled) {
+  if (!is_jit_enabled) {
+    return;
+  }
+
+  auto graph = node->owningGraph();
+
+  WithInsertPoint insert_point(node);
+
+  Value* true_constant = graph->insertConstant(IValue(true));
+  node->output()->replaceAllUsesWith(true_constant);
+  node->destroy();
```
@eellison would there be any concern here that this makes debugging hard? This will entirely remove the `aten::is_autocast_enabled` node & replace it with a constant

eellison replied:
Don't think it's that much different than the pass that specializes the value.

eellison commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -649,6 +650,10 @@ static const std::vector<OperatorGeneratorArgs> opGenArgs{
         },
         aliasAnalysisFromSchema()),
+    OperatorGeneratorArgs(
+        TORCH_SELECTIVE_SCHEMA("aten::is_autocast_enabled() -> bool"),
```
We don't need an entry. You only need to add to interned_strings if you want to use the Symbol in a switch statement or without the `Symbol::aten("is_autocast_enabled")`.  (it's fine to add it, but it's not strictly needed).

eellison reviewed (approved):
LGTM, one comment

davidberard98 commented on torch/csrc/jit/passes/autocast.cpp:
```diff
@@ -216,6 +216,39 @@ void castInputsToWidestType(Node* node, const AutocastContext& context) {
+// invoked, then autocasting will occur regardless of what the JIT-autocasting
+// settings are.
+void updateAutocastEnabledCheck(Node* node, bool is_jit_enabled) {
+  if (!is_jit_enabled) {
```
@eellison If we do:
```
with torch.cuda.amp.autocast(True):
    scripted_fn(x, y, z, ...)
```
then autocasting will still happen, even if autocasting has been disabled by a scripted context manager.

davidberard98 committed:
Update on "[JIT] Support scripting torch.is_autocast_enabled()"


This adds an `aten::is_autocast_enabled` op into the jit runtime so that
autocasting ops can be scripted and called from within jit.

Differential Revision: [D37901585](https://our.internmc.facebook.com/intern/diff/D37901585)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37901585).

davidberard98 commented:
@pytorchbot merge

xsacha commented:
Prior to this commit, I can enable AMP and then freeze the model, which will bake in autocast in to the model.

With this change, I would now need to bake in the is_autocast and then enable AMP during a forward pass?

malfet commented:
@pytorchbot revert -m "Broke lite-intepreter builds, see https://github.com/pytorch/pytorch/runs/7550084494?check_suite_focus=true" -c nosignal
